packages/context/test/unit/context.unit.ts

```diff
@@ -681,6 +685,109 @@ describe('Context', () => {
+      expect(ctx.isSubscribed(nonMatchingListener)).to.true();
+      ctx.unsubscribe(nonMatchingListener);
+      expect(ctx.isSubscribed(nonMatchingListener)).to.false();
+    });
+
+    it('allows subscription.unsubscribe()', () => {
+      const subscription = ctx.subscribe(nonMatchingListener);
+      expect(ctx.isSubscribed(nonMatchingListener)).to.true();
+      subscription.unsubscribe();
+      expect(ctx.isSubscribed(nonMatchingListener)).to.false();
+      expect(subscription.closed).to.be.true();
+    });
+
+    it('registers listeners on context chain', () => {
+      const childCtx = new Context(ctx, 'child');
+      childCtx.subscribe(nonMatchingListener);
+      expect(childCtx.isSubscribed(nonMatchingListener)).to.true();
+      expect(ctx.isSubscribed(nonMatchingListener)).to.true();
+    });
+
+    it('un-registers listeners on context chain', () => {
+      const childCtx = new Context(ctx, 'child');
+      childCtx.subscribe(nonMatchingListener);
+      expect(childCtx.isSubscribed(nonMatchingListener)).to.true();
+      expect(ctx.isSubscribed(nonMatchingListener)).to.true();
+      childCtx.unsubscribe(nonMatchingListener);
+      expect(childCtx.isSubscribed(nonMatchingListener)).to.false();
+      expect(ctx.isSubscribed(nonMatchingListener)).to.false();
+    });
+
+    function givenNonMatchingListener() {
+      nonMatchingListener = {
+        filter: binding => false,
+        listen: (event, binding) => {},
+      };
+    }
+  });
+
+  describe('event notification', () => {
+    let matchingListener: ContextEventListener;
+    let nonMatchingListener: ContextEventListener;
+    const events: string[] = [];
+    let nonMatchingListenerCalled = false;
+
+    beforeEach(givenListeners);
+
+    it('emits bind event to matching listeners', async () => {
+      ctx.bind('foo').to('foo');
+      await nextTick();
+      expect(events).to.eql(['foo:bind']);
+      expect(nonMatchingListenerCalled).to.be.false();
+    });
+
+    it('emits unbind event to matching listeners', async () => {
+      ctx.bind('foo').to('foo');
+      await nextTick();
+      ctx.unbind('foo');
+      await nextTick();
+      expect(events).to.eql(['foo:bind', 'foo:unbind']);
+      expect(nonMatchingListenerCalled).to.be.false();
+    });
+
+    it('does not trigger listeners if affected binding is the same', async () => {
+      const binding = ctx.bind('foo').to('foo');
+      await nextTick();
+      expect(events).to.eql(['foo:bind']);
+      ctx.add(binding);
+      await nextTick();
+      expect(events).to.eql(['foo:bind']);
+    });
+
+    function givenListeners() {
+      nonMatchingListenerCalled = false;
+      events.splice(0, events.length);
+      nonMatchingListener = {
+        filter: binding => false,
+        listen: (event, binding) => {
+          nonMatchingListenerCalled = true;
+        },
+      };
+      matchingListener = {
+        filter: binding => true,
+        listen: (event, binding) => {
```

Comment:
Please add and verify how async listeners are handled too.

```ts
asyncMatchingListener = {
  filter: binding => true,
  listen: async (event, binding) => { 
    await nextTick(); 
    events.push(`ASYNC:${binding.key}:${event}`); 
  },
};
```

Besides the basic scenarios shown in the existing tests, please cover the following more complex situation too.

There are two listeners (sync first, async second) registered with the context. Now we add two bindings "A" and "B". In which order are notifications delivered?

I believe your current implementation will do the following:
 - tick 0: bindings are created
 - tick 1: listeners are notified about both events. The first (sync) listeners pushes both A and B notifications to `events`. The second listener (async) is waiting for the next tick.
 - tick 2: The async listener handles notifications, and pushes A and B notifications to `events`.
Outcome: `['A', 'B', 'ASYNC:A', 'ASYNC:B']`.

I think the following order of calls will less surprising to LB4 users?
- tick 0: bindings are created
- tick 1: notification about A is processed. The sync listener pushes notification `A` to `events`. The async listeners starts the wait.
- tick 2: The async listener processes notification about A and pushes an entry to `events`. The sync listener receives notification about B and pushes an entry to `events`. The async listener receives notification about B and starts the wait.
- tick 3: The async listener pushes `B` to `events`.
Outcome: `['A', 'ASYNC:A', 'B', 'ASYNC:B']`.

Let's discuss.

Reply:
Added.